# Fake News Implementation

In [13]:
import pandas as pd

df=pd.read_csv('train.csv')
df.head(n=3)

id                                              title              author  \
0   0  House Dem Aide: We Didn’t Even See Comey’s Let...       Darrell Lucus   
1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...     Daniel J. Flynn   
2   2                  Why the Truth Might Get You Fired  Consortiumnews.com   

                                                text  label  
0  House Dem Aide: We Didn’t Even See Comey’s Let...      1  
1  Ever get the feeling your life circles the rou...      0  
2  Why the Truth Might Get You Fired October 29, ...      1

In [14]:
## Get the Independent Features
X=df.drop('label',axis=1)

In [15]:
X.head(n=3)

id                                              title              author  \
0   0  House Dem Aide: We Didn’t Even See Comey’s Let...       Darrell Lucus   
1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...     Daniel J. Flynn   
2   2                  Why the Truth Might Get You Fired  Consortiumnews.com   

                                                text  
0  House Dem Aide: We Didn’t Even See Comey’s Let...  
1  Ever get the feeling your life circles the rou...  
2  Why the Truth Might Get You Fired October 29, ...

In [16]:
## Get the Dependent features
y=df['label']
y.head(n=5)

0    1
1    0
2    1
3    1
4    1
Name: label, dtype: int64

In [17]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, HashingVectorizer

### As we don't want to apply any feature enginnering on NaN values  so we will drop them by using dropna()

In [18]:
# Droping the NULL values

print(df.shape)
df=df.dropna()
print(df.shape)

(20800, 5)
(18285, 5)


In [19]:
# Reseting the index as many index values will get droped

messages=df.copy()
messages.reset_index(inplace=True)
messages.head(n=7)

index  id                                              title  \
0      0   0  House Dem Aide: We Didn’t Even See Comey’s Let...   
1      1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...   
2      2   2                  Why the Truth Might Get You Fired   
3      3   3  15 Civilians Killed In Single US Airstrike Hav...   
4      4   4  Iranian woman jailed for fictional unpublished...   
5      5   5  Jackie Mason: Hollywood Would Love Trump if He...   
6      7   7  Benoît Hamon Wins French Socialist Party’s Pre...   

               author                                               text  \
0       Darrell Lucus  House Dem Aide: We Didn’t Even See Comey’s Let...   
1     Daniel J. Flynn  Ever get the feeling your life circles the rou...   
2  Consortiumnews.com  Why the Truth Might Get You Fired October 29, ...   
3     Jessica Purkiss  Videos 15 Civilians Killed In Single US Airstr...   
4      Howard Portnoy  Print \nAn Iranian woman has been sentenced to...   
5     Daniel Nussbaum  In these trying times, Jackie Mason is the Voi...   
6     Alissa J. Rubin  PARIS  —   France chose an idealistic, traditi...   

   label  
0      1  
1      0  
2      1  
3      1  
4      1  
5      0  
6      0

## We will Process the text segment of the table as it is relatively larger segment than title segment

## This Block will take time to execute as there are 18285 text paragraph

In [ ]:
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
import re
ps = PorterStemmer()
corpus = []


for i in range(0, len(messages)):
    # substitue all the characters other than alphabets with blank
    review = re.sub('[^a-zA-Z]', ' ', messages['text'][i])
    review = review.lower()
    review = review.split()
    
    review = [ps.stem(word) for word in review if not word in stopwords.words('english')]
    review = ' '.join(review)
    corpus.append(review)

In [ ]:
corpus[3]

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer


# from all the words take most frequent 5000 words
# ngram_range means take the combination of 1 word 2 word or 3 word as a feature and then apply BOG
# i.e say combination of 1 ,2 or 3 words may be vigilent in feature names i.e say "ABP news" might be the single 
# feature which itself are two words

tfidf_v=TfidfVectorizer(max_features=5000,ngram_range=(1,3))
X=tfidf_v.fit_transform(corpus).toarray()

In [ ]:
X.shape

In [ ]:
y=messages['label']